File: archived/v1/src/c/ch01/1.3/l1/fractions_utils.c
#include "fractions_utils.h"
#include "gcd_lcm.h"

frac64 reduce_fraction64(int64_t n, int64_t d) {
    /* caller must ensure d != 0 */
    int64_t g = gcd64(n, d);
    int64_t num = n / g;
    int64_t den = d / g;
    if (den < 0) { num = -num; den = -den; }
    return (frac64){ .num = num, .den = den };
}

File: archived/v1/src/erlang/ch01/1.3/l1/src/fractions_utils.erl
-module(fractions_utils).
-export([reduce_fraction/1]).

%% reduce_fraction({N,D}) -> {Num, Den} in lowest terms with Den > 0.
reduce_fraction({_N, 0}) ->
    erlang:error(badarg);
reduce_fraction({N, D}) ->
    G = gcd_lcm:gcd(N, D),
    Num0 = N div G,
    Den0 = D div G,
    case Den0 < 0 of
        true  -> {-Num0, -Den0};
        false -> {Num0, Den0}
    end.